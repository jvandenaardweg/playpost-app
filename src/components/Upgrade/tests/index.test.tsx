import React from 'react';
import { fireEvent, render, RenderAPI } from 'react-native-testing-library';

import { SUBSCRIPTION_PRODUCT_ID_FREE, SUBSCRIPTION_PRODUCT_ID_FREE_ANDROID, SUBSCRIPTION_PRODUCT_ID_FREE_APPLE, SUBSCRIPTION_PRODUCT_ID_PLUS, SUBSCRIPTION_PRODUCT_ID_PLUS_ANDROID, SUBSCRIPTION_PRODUCT_ID_PLUS_APPLE, SUBSCRIPTION_PRODUCT_ID_PREMIUM, SUBSCRIPTION_PRODUCT_ID_PREMIUM_ANDROID, SUBSCRIPTION_PRODUCT_ID_PREMIUM_APPLE } from '../../../constants/in-app-purchase';
import { Upgrade } from '../index';

import mockSubscriptionsAndroid from '../../../../tests/__mocks__/react-native-iap/subscriptions-android';
import mockSubscriptionsIOS from '../../../../tests/__mocks__/react-native-iap/subscriptions-ios';

const subscriptionFeaturesIOS = [
  {
    productId: SUBSCRIPTION_PRODUCT_ID_FREE,
    title: 'Free',
    price: '0',
    body: ['Basic quality voices', 'One voice option per language', 'Max. 30 minutes per month', 'Unlimited playlist items', 'Some advertisements'],
    footer: 'About 5 articles to audio, per month'
  },
  {
    productId: SUBSCRIPTION_PRODUCT_ID_PREMIUM,
    title: 'Premium',
    price: null,
    body: [`10+ High Quality voices`, 'Multiple voice options per language', 'Max. 120 minutes per month', 'Unlimited playlist items', 'No advertisements'],
    footer: 'About 25 articles to audio, per month'
  },
  {
    productId: SUBSCRIPTION_PRODUCT_ID_PLUS,
    title: 'Plus',
    price: null,
    body: [`10+ High Quality voices`, 'Multiple voice options per language', 'Max. 300 minutes per month', 'Unlimited playlist items', 'No advertisements'],
    footer: 'About 65 articles to audio, per month'
  }
]

const subscriptionFeaturesAndroid = [
  {
    productId: SUBSCRIPTION_PRODUCT_ID_FREE_ANDROID,
    title: 'Free',
    price: '0',
    body: ['Basic quality voices', 'One voice option per language', 'Max. 30 minutes per month', 'Unlimited playlist items', 'Some advertisements'],
    footer: 'About 5 articles to audio, per month'
  },
  {
    productId: SUBSCRIPTION_PRODUCT_ID_PREMIUM_ANDROID,
    title: 'Premium',
    price: null,
    body: [`10+ High Quality voices`, 'Multiple voice options per language', 'Max. 120 minutes per month', 'Unlimited playlist items', 'No advertisements'],
    footer: 'About 25 articles to audio, per month'
  },
  {
    productId: SUBSCRIPTION_PRODUCT_ID_PLUS_ANDROID,
    title: 'Plus',
    price: null,
    body: [`10+ High Quality voices`, 'Multiple voice options per language', 'Max. 300 minutes per month', 'Unlimited playlist items', 'No advertisements'],
    footer: 'About 65 articles to audio, per month'
  }
]

const onPressUpgradeHandler = jest.fn();
const onPressRestoreHandler = jest.fn();
const onPressPrivacyHandler = jest.fn();
const onPressTermsHandler = jest.fn();
const onPressCancelHandler = jest.fn();
const isDowngradePaidSubscriptionHandler = jest.fn();

// TODO: use Prop type when RNIap types are correct
const defaultProps: any = {
  isLoadingSubscriptionItems: false,
  isLoadingBuySubscription: false,
  isLoadingRestorePurchases: false,
  isEligibleForTrial: true,
  subscriptions: mockSubscriptionsIOS,
  activeSubscriptionProductId: SUBSCRIPTION_PRODUCT_ID_FREE,
  centeredSubscriptionProductId: SUBSCRIPTION_PRODUCT_ID_PREMIUM,
  subscriptionFeatures: subscriptionFeaturesIOS,
  onPressUpgrade: onPressUpgradeHandler,
  onPressRestore: onPressRestoreHandler,
  onPressPrivacy: onPressPrivacyHandler,
  onPressTerms: onPressTermsHandler,
  onPressCancel: onPressCancelHandler,
  isDowngradePaidSubscription: isDowngradePaidSubscriptionHandler,
}

describe('Upgrade', () => {
  describe('rendering', () => {
    let wrapper: RenderAPI;

    afterEach(() => {
      jest.resetAllMocks();
      jest.resetModules();
    });

    beforeAll(() => {
      const props = {
        ...defaultProps
      }

      wrapper = render(<Upgrade {...props} />);
    });

    it('should render correctly', () => {
      expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('should render the correct subscription information for iOS', () => {
      const Platform = require('react-native').Platform;
      Platform.OS = 'ios';

      const props = {
        ...defaultProps,
        subscriptions: mockSubscriptionsIOS,
        isLoadingSubscriptionItems: false,
        isEligibleForTrial: true,
        subscriptionFeatures: subscriptionFeaturesIOS
      }

      wrapper.update(<Upgrade {...props} />)

      expect(wrapper.getAllByTestId('Upgrade-Text-price')[0].props.children).toBe('€0');
      expect(wrapper.getAllByTestId('Upgrade-Text-price')[1].props.children).toBe('€ 4,99');
      expect(wrapper.getAllByTestId('Upgrade-Text-price')[2].props.children).toBe('€ 9,99');

      expect(wrapper.getAllByTestId('Upgrade-Button')[0].props.children[0].props.children[2].props.children).toBe('Current subscription'); // free
      expect(wrapper.getAllByTestId('Upgrade-Button')[1].props.children[0].props.children[2].props.children).toBe('Start free 3-day trial'); // premium
      expect(wrapper.getAllByTestId('Upgrade-Button')[2].props.children[0].props.children[2].props.children).toBe('Start free 3-day trial'); // plus

    });

    it('should render the correct subscription information for Android', () => {
      const Platform = require('react-native').Platform;
      Platform.OS = 'android';

      const props = {
        ...defaultProps,
        subscriptions: mockSubscriptionsAndroid,
        isLoadingSubscriptionItems: false,
        isEligibleForTrial: true,
        subscriptionFeatures: subscriptionFeaturesAndroid
      }

      wrapper.update(<Upgrade {...props} />)

      expect(wrapper.getAllByTestId('Upgrade-Text-price')[0].props.children).toBe('€0');
      expect(wrapper.getAllByTestId('Upgrade-Text-price')[1].props.children).toBe('€4.99');
      expect(wrapper.getAllByTestId('Upgrade-Text-price')[2].props.children).toBe('€9.99');

      expect(wrapper.getAllByTestId('Upgrade-Button')[0].props.children[0].props.children[2].props.children).toBe('Current subscription'); // free
      expect(wrapper.getAllByTestId('Upgrade-Button')[1].props.children[0].props.children[2].props.children).toBe('Start free 3-day trial'); // premium
      expect(wrapper.getAllByTestId('Upgrade-Button')[2].props.children[0].props.children[2].props.children).toBe('Start free 3-day trial'); // plus

    });

    it('should render the correct active subscription for iOS', () => {
      const Platform = require('react-native').Platform;
      Platform.OS = 'ios';

      const props = {
        ...defaultProps,
        subscriptions: mockSubscriptionsIOS,
        isLoadingSubscriptionItems: false,
        isEligibleForTrial: false,
        subscriptionFeatures: subscriptionFeaturesIOS,
        activeSubscriptionProductId: SUBSCRIPTION_PRODUCT_ID_PREMIUM_APPLE,
        centeredSubscriptionProductId: SUBSCRIPTION_PRODUCT_ID_PREMIUM_APPLE,
      }

      wrapper.update(<Upgrade {...props} />)

      expect(wrapper.getAllByTestId('Upgrade-Button')[0].props.children[0].props.children[2].props.children).toBe('Downgrade to Free'); // free
      expect(wrapper.getAllByTestId('Upgrade-Button')[1].props.children[0].props.children[2].props.children).toBe('Current subscription'); // premium
      expect(wrapper.getAllByTestId('Upgrade-Button')[2].props.children[0].props.children[2].props.children).toBe('Upgrade to Plus'); // plus

    });

    it('should render the correct active subscription for Android', () => {
      const Platform = require('react-native').Platform;
      Platform.OS = 'android';

      const props = {
        ...defaultProps,
        subscriptions: mockSubscriptionsAndroid,
        isLoadingSubscriptionItems: false,
        isEligibleForTrial: false,
        subscriptionFeatures: subscriptionFeaturesAndroid,
        activeSubscriptionProductId: SUBSCRIPTION_PRODUCT_ID_PREMIUM_ANDROID,
        centeredSubscriptionProductId: SUBSCRIPTION_PRODUCT_ID_PREMIUM_ANDROID,
      }

      wrapper.update(<Upgrade {...props} />)

      expect(wrapper.getAllByTestId('Upgrade-Button')[0].props.children[0].props.children[2].props.children).toBe('Downgrade to Free'); // free
      expect(wrapper.getAllByTestId('Upgrade-Button')[1].props.children[0].props.children[2].props.children).toBe('Current subscription'); // premium
      expect(wrapper.getAllByTestId('Upgrade-Button')[2].props.children[0].props.children[2].props.children).toBe('Upgrade to Plus'); // plus

    });

    it('should fire onPressUpgrade when pressing on the upgrade button on Android', () => {
      const Platform = require('react-native').Platform;
      Platform.OS = 'android';

      const props = {
        ...defaultProps,
        subscriptions: mockSubscriptionsAndroid,
        isLoadingSubscriptionItems: false,
        isEligibleForTrial: false,
        subscriptionFeatures: subscriptionFeaturesAndroid,
        activeSubscriptionProductId: SUBSCRIPTION_PRODUCT_ID_PREMIUM_ANDROID,
        centeredSubscriptionProductId: SUBSCRIPTION_PRODUCT_ID_PREMIUM_ANDROID,
      }

      wrapper.update(<Upgrade {...props} />)

      fireEvent.press(wrapper.getAllByTestId('Upgrade-Button')[0]);
      fireEvent.press(wrapper.getAllByTestId('Upgrade-Button')[1]);
      fireEvent.press(wrapper.getAllByTestId('Upgrade-Button')[2]);

      expect(onPressUpgradeHandler).toHaveBeenCalledTimes(3);
      expect(onPressUpgradeHandler).toHaveBeenNthCalledWith(1, SUBSCRIPTION_PRODUCT_ID_FREE_ANDROID);
      expect(onPressUpgradeHandler).toHaveBeenNthCalledWith(2, SUBSCRIPTION_PRODUCT_ID_PREMIUM_ANDROID);
      expect(onPressUpgradeHandler).toHaveBeenNthCalledWith(3, SUBSCRIPTION_PRODUCT_ID_PLUS_ANDROID);

    });

    it('should fire onPressUpgrade when pressing on the upgrade button on iOS', () => {
      const Platform = require('react-native').Platform;
      Platform.OS = 'ios';

      const props = {
        ...defaultProps,
        subscriptions: mockSubscriptionsIOS,
        isLoadingSubscriptionItems: false,
        isEligibleForTrial: false,
        subscriptionFeatures: subscriptionFeaturesIOS,
        activeSubscriptionProductId: SUBSCRIPTION_PRODUCT_ID_PREMIUM_APPLE,
        centeredSubscriptionProductId: SUBSCRIPTION_PRODUCT_ID_PREMIUM_APPLE,
      }

      wrapper.update(<Upgrade {...props} />)

      fireEvent.press(wrapper.getAllByTestId('Upgrade-Button')[0]);
      fireEvent.press(wrapper.getAllByTestId('Upgrade-Button')[1]);
      fireEvent.press(wrapper.getAllByTestId('Upgrade-Button')[2]);

      expect(onPressUpgradeHandler).toHaveBeenCalledTimes(3);
      expect(onPressUpgradeHandler).toHaveBeenNthCalledWith(1, SUBSCRIPTION_PRODUCT_ID_FREE_APPLE);
      expect(onPressUpgradeHandler).toHaveBeenNthCalledWith(2, SUBSCRIPTION_PRODUCT_ID_PREMIUM_APPLE);
      expect(onPressUpgradeHandler).toHaveBeenNthCalledWith(3, SUBSCRIPTION_PRODUCT_ID_PLUS_APPLE);

    });
  });
});
