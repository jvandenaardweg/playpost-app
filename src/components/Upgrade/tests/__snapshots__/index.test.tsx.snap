// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upgrade rendering should render correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    collapsable={false}
    contentContainerStyle={
      Object {
        "padding": 16,
      }
    }
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
  >
    <View>
      <View
        style={
          Object {
            "marginBottom": 16,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "borderBottomWidth": 1,
              "borderColor": "rgba(0,0,0,0.05)",
              "flex": 1,
              "flexDirection": "row",
              "marginBottom": 14,
              "paddingBottom": 14,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 28,
                },
                Object {
                  "color": "#037DE2",
                  "marginLeft": 8,
                  "marginRight": 16,
                  "width": 40,
                },
                Object {
                  "fontFamily": "FontAwesome5Free-Regular",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {
                  "fontWeight": "400",
                },
              ]
            }
          >
            
          </Text>
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "column",
              }
            }
          >
            <Text
              style={
                Object {
                  "fontSize": 17,
                  "fontWeight": "600",
                  "lineHeight": 21,
                  "marginBottom": 4,
                }
              }
            >
              Higher quality voices
            </Text>
            <Text
              style={
                Object {
                  "color": "#888888",
                  "fontSize": 15,
                  "lineHeight": 24,
                }
              }
            >
              Access to our highest quality Premium voices for easier listening. You can preview these Premium voices in the settings screen.
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "borderBottomWidth": 1,
              "borderColor": "rgba(0,0,0,0.05)",
              "flex": 1,
              "flexDirection": "row",
              "marginBottom": 14,
              "paddingBottom": 14,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 28,
                },
                Object {
                  "color": "#037DE2",
                  "marginLeft": 8,
                  "marginRight": 16,
                  "width": 40,
                },
                Object {
                  "fontFamily": "FontAwesome5Free-Solid",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {
                  "fontWeight": "900",
                },
              ]
            }
          >
            
          </Text>
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "column",
              }
            }
          >
            <Text
              style={
                Object {
                  "fontSize": 17,
                  "fontWeight": "600",
                  "lineHeight": 21,
                  "marginBottom": 4,
                }
              }
            >
              Voice customization options
            </Text>
            <Text
              style={
                Object {
                  "color": "#888888",
                  "fontSize": 15,
                  "lineHeight": 24,
                }
              }
            >
              Choose between a variety of male and female voices with accents like American, British or Australian English.
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "borderBottomWidth": 1,
              "borderColor": "rgba(0,0,0,0.05)",
              "flex": 1,
              "flexDirection": "row",
              "marginBottom": 14,
              "paddingBottom": 14,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 28,
                },
                Object {
                  "color": "#037DE2",
                  "marginLeft": 8,
                  "marginRight": 16,
                  "width": 40,
                },
                Object {
                  "fontFamily": "FontAwesome5Free-Regular",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {
                  "fontWeight": "400",
                },
              ]
            }
          >
            
          </Text>
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "column",
              }
            }
          >
            <Text
              style={
                Object {
                  "fontSize": 17,
                  "fontWeight": "600",
                  "lineHeight": 21,
                  "marginBottom": 4,
                }
              }
            >
              Unlimited listening
            </Text>
            <Text
              style={
                Object {
                  "color": "#888888",
                  "fontSize": 15,
                  "lineHeight": 24,
                }
              }
            >
              No article time limits. Listen to articles longer than 5 minutes.
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "borderBottomWidth": 1,
              "borderColor": "rgba(0,0,0,0.05)",
              "flex": 1,
              "flexDirection": "row",
              "marginBottom": 14,
              "paddingBottom": 14,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 28,
                },
                Object {
                  "color": "#037DE2",
                  "marginLeft": 8,
                  "marginRight": 16,
                  "width": 40,
                },
                Object {
                  "fontFamily": "FontAwesome5Free-Solid",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {
                  "fontWeight": "900",
                },
              ]
            }
          >
            
          </Text>
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "column",
              }
            }
          >
            <Text
              style={
                Object {
                  "fontSize": 17,
                  "fontWeight": "600",
                  "lineHeight": 21,
                  "marginBottom": 4,
                }
              }
            >
              No advertisements
            </Text>
            <Text
              style={
                Object {
                  "color": "#888888",
                  "fontSize": 15,
                  "lineHeight": 24,
                }
              }
            >
              Sponsored content helps us continue to provide a free version of Playpost. After upgrading to Playpost Premium, you won’t see any ads, and you’ll be supporting Playpost more directly!
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "paddingBottom": 24,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 13,
              "lineHeight": 18,
              "opacity": 0.5,
            }
          }
        >
          Payment will be charged to your Apple ID account at the confirmation of purchase. Subscription automatically renews unless it is canceled at least 24 hours before the end of the current period. Your account will be charged for renewal within 24 hours prior to the end of the current period. You can manage and cancel your subscriptions by going to your account settings on the App Store after purchase.
        </Text>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "center",
              "marginTop": 6,
            }
          }
        >
          <Text
            onPress={[Function]}
            style={
              Array [
                Object {
                  "fontSize": 13,
                  "lineHeight": 18,
                  "opacity": 0.5,
                },
                Object {
                  "color": "#000000",
                  "opacity": 1,
                  "textDecorationLine": "underline",
                },
              ]
            }
          >
            Privacy Policy
          </Text>
          <Text
            style={
              Object {
                "fontSize": 13,
                "lineHeight": 18,
                "opacity": 0.5,
              }
            }
          >
             - 
          </Text>
          <Text
            onPress={[Function]}
            style={
              Array [
                Object {
                  "fontSize": 13,
                  "lineHeight": 18,
                  "opacity": 0.5,
                },
                Object {
                  "color": "#000000",
                  "opacity": 1,
                  "textDecorationLine": "underline",
                },
              ]
            }
          >
            Terms of Use
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "borderColor": "#EEEEEE",
        "borderTopWidth": 1,
        "padding": 24,
      }
    }
  >
    <View
      style={
        Object {
          "borderRadius": 3,
        }
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#2089dc",
              "borderColor": "#2089dc",
              "borderRadius": 3,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
                "paddingBottom": 1,
                "paddingTop": 2,
                "textAlign": "center",
              }
            }
          >
            Upgrade for 1 euro
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "borderRadius": 3,
        }
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "#2089dc",
              "borderRadius": 3,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#2089dc",
                "fontSize": 15,
                "paddingBottom": 1,
                "paddingTop": 2,
                "textAlign": "center",
              }
            }
          >
            Already upgraded? Restore purchase
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
